packages/plugins/opentelemetry/src/plugin.ts

```diff
@@ -0,0 +1,291 @@
+      execReqSubgraphName.set(executionRequest, subgraphName);
+    },
+    onRequest(onRequestPayload) {
+      const shouldTraceHttp =
+        typeof options.spans?.http === 'function'
+          ? options.spans.http(onRequestPayload)
+          : (options.spans?.http ?? true);
+
+      if (!shouldTraceHttp) {
+        return;
+      }
+
+      const { request, url } = onRequestPayload;
+      const rootContext = propagation.extract(context.active(), request.headers);
+      const path = url.pathname;
+      const userAgent = request.headers.get('user-agent');
+      const ips = request.headers.get('x-forwarded-for');
+      const method = request.method || 'GET';
+      const host = url.host || request.headers.get('host');
+      const hostname = url.hostname || host || 'localhost';
+
+      const rootSpanName = `${method} ${path}`;
+      const rootSpan = tracer.startSpan(
+        rootSpanName,
+        {
+          attributes: {
+            [SEMATTRS_HTTP_METHOD]: method,
+            [SEMATTRS_HTTP_URL]: request.url,
+            [SEMATTRS_HTTP_ROUTE]: path,
+            [SEMATTRS_HTTP_SCHEME]: url.protocol,
+            [SEMATTRS_NET_HOST_NAME]: hostname,
+            [SEMATTRS_HTTP_HOST]: host,
+            [SEMATTRS_HTTP_CLIENT_IP]: ips?.split(',')[0],
+            [SEMATTRS_HTTP_SERVER_NAME]: serviceName,
+            [SEMATTRS_HTTP_USER_AGENT]: userAgent,
+          },
+          kind: SpanKind.SERVER,
+        },
+        rootContext,
+      );
+
+      requestContextMapping.set(request, trace.setSpan(rootContext, rootSpan));
+    },
+    onParse(onParsePayload) {
+      const shouldTracePrase =
+        typeof options.spans?.graphqlParse === 'function'
+          ? options.spans.graphqlParse(onParsePayload)
+          : (options.spans?.graphqlParse ?? true);
+
+      const { context } = onParsePayload;
+
+      if (shouldTracePrase && requestContextMapping.has(context.request)) {
+        const rootContext = requestContextMapping.get(context.request)!;
+        const parseSpan = tracer.startSpan(
+          'graphql.parse',
+          {
+            attributes: {
+              [SEMATTRS_GRAPHQL_DOCUMENT]: context.params.query,
+              [SEMATTRS_GRAPHQL_OPERATION_NAME]: context.params.operationName,
+            },
+            kind: SpanKind.INTERNAL,
+          },
+          rootContext,
+        );
+
+        return ({ result }) => {
+          if (result instanceof Error) {
+            parseSpan.setStatus({
+              code: SpanStatusCode.ERROR,
+              message: result.message,
+            });
+          }
+
+          parseSpan.end();
+        };
+      }
+    },
+    onExecute({ args }) {
+      const shouldTraceExecute = options.spans?.graphqlExecute ?? true;
+      if (requestContextMapping.has(args.contextValue.request)) {
+        const rootContext = requestContextMapping.get(args.contextValue.request)!;
+        const operation = getOperationAST(args.document, args.operationName);
```

Comment:
```suggestion
        const operation = getOperationASTFromDocument(args.document, args.operationName);
```
`getOperationASTFromDocument` from '@graphql-tools/utils' allows memoization.

Reply:
fixed